### DISCUSSION TITLE
Complex query with optional parameters

### DISCUSSION THREAD

**Discussion by ghost** [vitaly-t/pg-promise#856] [Q&A] [answered]
Hi,

currently i am working on complex query used for pagination and filtering big group of people, the problem i face with this task is issue with creating one readable query with many optional parameters (for example if user picks a city i need to use aggregation operation), so far my solution to that problem is cutting query into smaller parts where i use bunch of ifs and create subqueries strings with pgp.as.format that are later connected  into one big query, as you can assume managing and debugging such a query is one big nightmare, are there any solutions for such a problem in pg-promise, the best solution would be something that could drop a line/operation if parameter is not provided.

Thanks in advance for any tips

**Comment by vitaly-t** [comment, marked as answer]
Keep your query in external file, and add dynamic parts to it as raw text, pre-formatted via `pgp.as.format`.

See:
 * [Query Files](https://github.com/vitaly-t/pg-promise#query-files)
 * [Raw Text](https://github.com/vitaly-t/pg-promise#raw-text)
